src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Date.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Text.Json
+{
+    internal static partial class JsonWriterHelper
+    {
+        public static void TrimDateTime(Span<byte> buffer, out int bytesWritten)
+        {
+            Trim(buffer, out bytesWritten);
+        }
+
+        public static void TrimDateTimeOffset(Span<byte> buffer, out int bytesWritten)
+        {
+            Trim(buffer, out bytesWritten);
+        }
+
+        //
+        // Trims roundtrippable DateTime(Offset) input.
+        // If the milliseconds part of the date is zero, we omit the fraction part of the date,
+        // else we write the fraction up to 7 decimal places with no trailing zeros. i.e. the output format is
+        // YYYY-MM-DDThh:mm:ss[.s]TZD where TZD = Z or +-hh:mm.
+        // e.g.
+        //   ---------------------------------
+        //   2017-06-12T05:30:45.768-07:00
+        //   2017-06-12T05:30:45.00768Z           (Z is short for "+00:00" but also distinguishes DateTimeKind.Utc from DateTimeKind.Local)
+        //   2017-06-12T05:30:45                  (interpreted as local time wrt to current time zone)
+        private static void Trim(Span<byte> buffer, out int bytesWritten)
+        {
+            // Assert buffer is the right length for:
+            // YYYY-MM-DDThh:mm:ss.fffffff (JsonConstants.MaximumFormatDateTimeLength)
+            // YYYY-MM-DDThh:mm:ss.fffffffZ (JsonConstants.MaximumFormatDateTimeLength + 1)
+            // YYYY-MM-DDThh:mm:ss.fffffff(+|-)hh:mm (JsonConstants.MaximumFormatDateTimeOffsetLength)
+            Debug.Assert(buffer.Length == JsonConstants.MaximumFormatDateTimeLength ||
+                buffer.Length == (JsonConstants.MaximumFormatDateTimeLength + 1) ||
+                buffer.Length == JsonConstants.MaximumFormatDateTimeOffsetLength);
+
+            uint digit7 = buffer[26] - (uint)'0';
+            uint digit6 = buffer[25] - (uint)'0';
+            uint digit5 = buffer[24] - (uint)'0';
+            uint digit4 = buffer[23] - (uint)'0';
+            uint digit3 = buffer[22] - (uint)'0';
+            uint digit2 = buffer[21] - (uint)'0';
+            uint digit1 = buffer[20] - (uint)'0';
+            uint fraction = (digit1 * 1_000_000) + (digit2 * 100_000) + (digit3 * 10_000) + (digit4 * 1_000) + (digit5 * 100) + (digit6 * 10) + digit7;
+
+            // The period's index
+            int curIndex = 19;
+
+            if (fraction > 0)
+            {
+                int numFractionDigits = 7;
+                // Remove trailing zeros
+                while (fraction % 10 == 0)
```

Comment:
Consider using `Math.DivRem` instead of % and div, if that's faster.

Reply:
For example [sharplab.io](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAGABJgRgG4BYAKEwGZ8AmXAYVwG9zc39qBLAOzlwFkCACh58AZgCdEAYzicA9twCUzVu3Wjc3AK6oAYlNkLuAEU4BzTnGS4AvLgDspMuvUB3ABacANgFNcQpIycoq4AKS4BHi29thKaq6qLomuQUahAPT2Uc4prjr6hiGmFlY2ALTZuSkAvgmumA5augbBxmaW1tVsdcls9VS4mvw0Iry4acUqLH2umgWt6SWdNvZO9e5efgFwEtq+8bPqM3kAkJoAjtrycr7j9vyIcB4AdGYAbgBKvqiBRcYwSLYQHybR8TQSH6IHgAE18EiU3TyQzEAUhqGh3DhElwAEJYodkewTkTXMBIYgANZIvK9UkTf6hexXG6cO5wGkpBaM5ZlXCVSKcnobdiNZqFNqKDplJG9GpAA==)

```text
C.M1(Int32)
    L0000: mov ecx, edx
    L0002: mov r8d, 0x7
    L0008: mov eax, 0x66666667
    L000d: imul ecx
    L000f: mov eax, edx
    L0011: shr eax, 0x1f
    L0014: sar edx, 0x2
    L0017: add edx, eax
    L0019: lea eax, [rdx+rdx*4]
    L001c: shl eax, 1
    L001e: mov edx, ecx
    L0020: sub edx, eax
    L0022: jnz L0054
    L0024: mov eax, 0x66666667
    L0029: imul ecx
    L002b: mov ecx, edx
    L002d: shr ecx, 0x1f
    L0030: sar edx, 0x2
    L0033: add ecx, edx
    L0035: dec r8d
    L0038: mov eax, 0x66666667
    L003d: imul ecx
    L003f: mov eax, edx
    L0041: shr eax, 0x1f
    L0044: sar edx, 0x2
    L0047: add edx, eax
    L0049: lea eax, [rdx+rdx*4]
    L004c: shl eax, 1
    L004e: mov edx, ecx
    L0050: sub edx, eax
    L0052: jz L0024
    L0054: mov eax, r8d
    L0057: ret

C.M2(Int32)
    L0000: mov ecx, edx
    L0002: mov r8d, 0x7
    L0008: mov eax, 0x66666667
    L000d: imul ecx
    L000f: mov eax, edx
    L0011: shr eax, 0x1f
    L0014: sar edx, 0x2
    L0017: add eax, edx
    L0019: lea edx, [rax+rax*4]
    L001c: shl edx, 1
    L001e: sub ecx, edx
    L0020: test ecx, ecx
    L0022: jnz L002b
    L0024: mov ecx, eax
    L0026: dec r8d
    L0029: jmp L0008
    L002b: mov eax, r8d
    L002e: ret
```

Reply:
```
BenchmarkDotNet=v0.11.4, OS=macOS Mojave 10.14.3 (18D109) [Darwin 18.2.0]
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.2.104
  [Host]     : .NET Core 2.2.2 (CoreCLR 4.6.27317.07, CoreFX 4.6.27318.02), 64bit RyuJIT
  DefaultJob : .NET Core 2.2.2 (CoreCLR 4.6.27317.07, CoreFX 4.6.27318.02), 64bit RyuJIT
```

I found `Math.DivRem` faster with `int`s:

```
| Method |      Mean |     Error |    StdDev | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------- |----------:|----------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
|    Mod | 12.133 ns | 0.2707 ns | 0.4597 ns |  1.00 |    0.00 |           - |           - |           - |                   - |
| DivRem |  8.816 ns | 0.1996 ns | 0.1867 ns |  0.72 |    0.03 |           - |           - |           - |                   - |
```

```C#
using System;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Diagnosers;
using BenchmarkDotNet.Running;

namespace divRem
{
    class Program
    {
        static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);
    }

    [MemoryDiagnoser]
    public class DivRemBenchmark
    {
        [Benchmark(Baseline = true)]
        public void Mod()
        {
            int fraction = 4000000;
            int numFractionDigits = 7;

            while (fraction % 10 == 0)
            {
                fraction /= 10;
                numFractionDigits -= 1;
            }
        }

        [Benchmark]
        public void DivRem()                
        {
            int fraction = 4000000;
            int numFractionDigits = 7;

            while (true)
            {
                int quotient = Math.DivRem(fraction, 10, out int remainder);
                if (remainder != 0)
                {
                    break;
                }
                fraction = quotient;
                numFractionDigits -= 1;
            }
        }
    }
}
```

But with input data (`fraction`) left as `uint`, using `%` and `/` is faster:

```
| Method |     Mean |     Error |    StdDev | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|------- |---------:|----------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
|    Mod | 6.908 ns | 0.1683 ns | 0.3201 ns |  1.00 |    0.00 |           - |           - |           - |                   - |
| DivRem | 8.552 ns | 0.2039 ns | 0.3625 ns |  1.24 |    0.09 |           - |           - |           - |                   - |
```

```C#
using System;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Diagnosers;
using BenchmarkDotNet.Running;

namespace divRem
{
    class Program
    {
        static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);
    }

    [MemoryDiagnoser]
    public class DivRemBenchmark
    {
        [Benchmark(Baseline = true)]
        public void Mod()
        {
            uint fraction = 4000000;
            int numFractionDigits = 7;

            while (fraction % 10 == 0)
            {
                fraction /= 10;
                numFractionDigits -= 1;
            }
        }

        [Benchmark]
        public void DivRem()                
        {
            uint fraction = 4000000;
            int numFractionDigits = 7;

            while (true)
            {
                uint quotient = (uint)Math.DivRem((int)fraction, 10, out int remainder);
                if (remainder != 0)
                {
                    break;
                }
                fraction = quotient;
                numFractionDigits -= 1;
            }
        }
    }
}
```